exploration/selection-matching-options.md

```diff
@@ -0,0 +1,547 @@
+
+Many programming languages feature multi-value selection using logic constructs such as Java's `switch` statement. These statements do depend on the order of the "cases":
+
+```java
+switch (count) {
+   case 0:
+   case 1:
+   case 2:
+   default:
+};
+```
+
+Programming constructs like `switch` generally match a single value exactly. Sometimes developers use features, such as "fall through", to allow two or more blocks of code to operate on a single value. But the core of a `switch` is binary matching of the value. Generally the same input (`count` in the example above) can't match more than one `case`.
+
+We could model `match` as a switch:
+
+```
+match {$count}
+when 1   {This is your last chance}
+when one {You have {$count} chance remaining}
+when *   {...}
+```
+vs.
+```
+match {$count}
+when one {You have {$count} chance remaining}
+when 1   {This message cannot be reached in English, but would be reached in Japanese!!}
+when *   {...}
+```
+
+This exposes developers and translators to managing the complexity versus having the API take care of it.
+
+
+#### Are there other complex matching cases? Or is `plural` everything?
+
+Currently there are no other complex rule-based selectors in ICU. However, there are a number of cases where complex matching might come into play. The criteria for it being a complex match are:
+
+1. The selector generates different output by locale (the way that plural generates `few` and `many` for e.g. Polish or fails to generate `one` in Japanese). If the states vary according to something other than locale all of the states have to be accounted for in the `root` locale resource and all translations (so it's not "complex" any more); or:
+2. The selector can generate more than one match for the same value (compare with plural `1` and `one`), with varying quality of match.
+3. Any combination of the above.
+
+The key thing here is that the static text produced by the translator needs to reflect the grammatical needs of the language and depends on knowing something about the (invisible) value being inserted at runtime.
+
+Some potential examples of (1) (and this is "thinking out loud"):
+
+1. **Date/time based selection.** Date/time types, including the newer Temporal types, can present complex matching needs. While _incremental time_ values (such as `java.time.Instant`, `java.util.Date`, or JavaScript's `Date`) can resolve every field and be cast to any time zone, Other types, such as `java.time.ZonedDate`, are incomplete. There are different calendars that can affect presentation and selection as well. Some cases for complex time selection include:
+   * **Relative time formats.** The values available (such as `yesterday`, `tomorrow`, `day after tomorrow`) vary by locale. Here's one example in the [German CLDR charts](https://unicode-org.github.io/cldr-staging/charts/latest/summary/de.html#1d45310cbcf1b2e5)
```

Comment:
Why is it important that the values very by locale? I assume the issue is that the value of `1` matches both `tomorrow` and `in 1 day`, but that seems to be the manifestation of the second criteria above? ("The selector can generate more than one match for the same value.")

Reply:
This is *also* a manifestation of (1). English has words for `tomorrow` and `yesterday`, but used counted phrases (`in x days`, `x days ago`) for other values. German (and some other languages) have _words_ for two days ago/from now, e.g. `vorgestern`. So the relative time format used varies by locale.

Reply:
No need to look beyond the worlds of English even to find "ereyesterday" and "overmorrow". :)